gateway/res_cache.go

```diff
@@ -0,0 +1,152 @@
+package gateway
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/TykTechnologies/tyk/storage"
+	"github.com/TykTechnologies/tyk/user"
+)
+
+const (
+	upstreamCacheHeader    = "x-tyk-cache-action-set"
+	upstreamCacheTTLHeader = "x-tyk-cache-action-set-ttl"
+)
+
+// ResponseCacheMiddleware is a caching middleware that will pull data from Redis instead of the upstream proxy
+type ResponseCacheMiddleware struct {
+	spec  *APISpec
+	store storage.Handler
+}
+
+func (m *ResponseCacheMiddleware) Name() string {
+	return "ResponseCacheMiddleware"
+}
+
+func (h *ResponseCacheMiddleware) Init(c interface{}, spec *APISpec) error {
+	h.spec = spec
+	return nil
+}
+
+func (h *ResponseCacheMiddleware) HandleError(rw http.ResponseWriter, req *http.Request) {
+}
+
+func (m *ResponseCacheMiddleware) EnabledForSpec() bool {
+	return m.spec.CacheOptions.EnableCache
+}
+
+func (m *ResponseCacheMiddleware) getTimeTTL(cacheTTL int64) int64 {
+	timeNow := time.Now().Unix()
+	return timeNow + cacheTTL
+}
+
+func (m *ResponseCacheMiddleware) encodePayload(payload string, timestamp int64) string {
+	sEnc := base64.StdEncoding.EncodeToString([]byte(payload))
+	return sEnc + "|" + fmt.Sprint(timestamp)
+}
+
+func (m *ResponseCacheMiddleware) Logger() *logrus.Entry {
+	return log.WithField("mw", m.Name())
+}
+
+// HandleResponse checks if the http.Response argument can be cached and caches it for future requests.
+func (m *ResponseCacheMiddleware) HandleResponse(w http.ResponseWriter, res *http.Response, r *http.Request, ses *user.SessionState) error {
+	// No cache of empty responses
+	if res == nil {
+		m.Logger().Warning("Upstream request must have failed, response is empty")
+		return nil
+	}
+
+	// Skip caching the request if cache disabled
+	if !m.EnabledForSpec() {
+		return nil
+	}
+
+	// Has cache been enabled on the request?
+	options := ctxGetCacheOptions(r)
+	if options == nil {
+		m.Logger().Debug("Request is not cacheable")
+		return nil
+	}
+
+	cacheThisRequest := true
+	cacheTTL := m.spec.CacheOptions.CacheTimeout
+
+	// make sure the status codes match if specified
+	if len(options.cacheOnlyResponseCodes) > 0 {
```

Comment:
I woulder why you decided to read `cacheOnlyResponseCodes` from options, from context.
Does not feel like it is dynamic variable, and it should be read right from api definition. E.g. just move following code here in response middleware: https://github.com/TykTechnologies/tyk/pull/4397/files#diff-6266e0dbd16cef89e6de86a2c893114ba07799c804e2138172f9f94b08cdded8R190-R194

Reply:
In testing I came across scenarios where the cache reader said this is a cacheable path, but the cache writer would say that it's not. Rather than duplicate the checks inline, the result of that logic is now in a single place, and carried over.

The spec issue should be visible around HEAD~3.